### PR TITLE
[SPARK-33584][SQL] Hive Metastore support filter by string type with date values

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr make Hive Metastore support filter by string type with date values, a common case is:
```scala
spark.sql("create table t1(id string) partitioned by (part string) stored as parquet")
spark.sql("insert into t1 values('1', '2019-01-01')")
spark.sql("insert into t1 values('2', '2019-01-02')")
spark.sql("select * from t1 where  part = date '2019-01-01' ").show
```

By [default](https://github.com/apache/spark/blob/6d31daeb6a2c5607ffe3b23ffb381626ad57f576/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala#L126-L129), we use `DateType` to compare `StringType` and `DateType`.

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 1]
```diff
@@ -757,6 +757,7 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
           case attr: Attribute => Some(attr)
           case Cast(child @ IntegralType(), dt: IntegralType, _)
               if Cast.canUpCast(child.dataType.asInstanceOf[AtomicType], dt) => unapply(child)
+          case Cast(child @ StringType(), DateType, _) => unapply(child)
```
Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:760]: What if an input str is a special value (e.g., now, today, ...)?

**Comment by HyukjinKwon** [comment]
cc @MaxGekk FYI

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 1 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:760]: These special values handle by:

https://github.com/apache/spark/blob/6d625ccd5b5a76a149e2070df31984610629a295/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/finishAnalysis.scala#L82-L87

https://github.com/apache/spark/blob/a5e13acd19871831a93a5bdcbc99a9eb9f1aba07/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala#L660-L661

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala [hunk 2]
```diff
@@ -108,6 +108,21 @@ class FiltersSuite extends SparkFunSuite with Logging with PlanTest {
     (a("datecol", DateType) =!= Literal(Date.valueOf("2019-01-01"))) :: Nil,
     "datecol != 2019-01-01")
 
+  filterTest("string filter with date type",
+    (a("strcol", StringType).cast(DateType) > Literal(Date.valueOf("2019-01-01"))) :: Nil,
+    "strcol > 2019-01-01")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:113]: is it really safe? Seems it assumes hive will also cast the string col to date and compare with the date literal.

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/client/FiltersSuite.scala:113]: is it really safe? Seems it assumes hive will also cast the string col to date and compare with the date literal.

**Pull request closed without merging**